modules/snmp-dest/snmpdest.c

```diff
@@ -0,0 +1,899 @@
+    {
+      if (!snmpdest_dd_session_init(self))
+        {
+          msg_warning("SNMP: error in session init, message dropped",
+                      evt_tag_str("host", self->host),
+                      evt_tag_int("port", self->port),
+                      NULL);
+          log_msg_unref(msg);
+          return FALSE;
+        }
+    }
+
+  pdu = snmp_pdu_create(SNMP_MSG_TRAP2);
+  if (!pdu)
+    return FALSE;
+
+  gchar **oid_tokens;
+  oid parsed_oids[MAX_OIDS];
+  guint oid_cnt;
+  GList *snmp_obj = self->snmp_objs;
+  GList *snmp_code = self->snmp_codes;
+  GList *snmp_template = self->snmp_templates;
+  GString *fs = g_string_sized_new(128);
+  /* go through the snmp objects and add them to the message */
+  while (snmp_obj)
+    {
+      gchar *oid = snmp_obj->data;
+      if (oid[0] == '.')
+        {
+          /* skip the leading '.' in order not to parse an empty first token */
+          ++oid;
+        }
+
+      /* parse the oid string into oid values */
+      oid_tokens = g_strsplit(oid, ".", MAX_OIDS);
+      oid_cnt = 0;
+      while (oid_tokens[oid_cnt])
+        {
+          gint val;
+          sscanf(oid_tokens[oid_cnt], "%d", &val);
+          parsed_oids[oid_cnt] = val;
+          oid_cnt += 1;
+        }
+
+      g_strfreev(oid_tokens);
+
+      log_template_format(snmp_template->data, msg, &self->template_options, LTZ_LOCAL, 0, NULL, fs);
+
+      gint code = *((gint *)snmp_code->data);
+      /* check the number format and replace it with zero if necessary */
+      if (code == 0 || code == 1 || code == 3)
+        {
+          gboolean replace = FALSE;
+          gint i,
+               len = strlen(fs->str);
+          if (!len)
+            replace = TRUE;
+          else
+            for (i = 0; i < len && !replace; ++i)
+              if (fs->str[i] < '0' || fs->str[i] > '9')
+                replace = TRUE;
+          if (replace)
+            {
+              msg_warning("SNMP: invalid number replaced with '0'",
+                          evt_tag_str("value", fs->str),
+                          NULL);
+              g_string_assign(fs, "0");
+            }
+        }
+
+      gchar type_code = object_type_codes[code];
+      if (snmp_add_var(pdu, parsed_oids, oid_cnt, type_code, fs->str) != 0)
+        {
+          msg_warning("SNMP: error adding variable",
+                      evt_tag_str("value", fs->str),
+                      NULL);
+          log_msg_unref(msg);
+          return FALSE;
+        }
+
+      snmp_obj = snmp_obj->next->next->next; /* go to the next triplet and the corresponding items */
+      snmp_code = snmp_code->next;
+      snmp_template = snmp_template->next;
+    }
+  g_string_free(fs, TRUE);
+
+  gint success = snmp_send(self->ss, pdu);
+  if (success == 0)
+    {
+      msg_error("SNMP: send error",
+                evt_tag_int("code", snmp_errno),
+                evt_tag_str("message", snmp_api_errstring(snmp_errno)),
+                NULL);
+      snmp_free_pdu(pdu);
```

Comment:
In the next line we acknowledge the message: the message is lost. This is probably worth addressing, but for that it would need to turn this into a threaded destination driver, which might be much work.

Another thing, does the dropped stats counter increase in this case?

Reply:
stats_counter_inc(self->dropped_messages) added